pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go

```diff
@@ -0,0 +1,193 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	udpa "github.com/cncf/udpa/go/udpa/type/v1"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	xdslistener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	hcm_filter "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"google.golang.org/protobuf/proto"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	securitymodel "istio.io/istio/pilot/pkg/security/model"
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultRuntime = "envoy.wasm.runtime.v8"
+
+	IstioStatsPluginName      = "istio.stats"
+	EnvoyHTTPRouterPluginName = "envoy.filters.http.router"
+
+	WasmFilterTypeURL = "envoy.extensions.filters.http.wasm.v3.Wasm"
+)
+
+var (
+	// CacheCluster is the Envoy cluster that is used to retrieve WASM filters from
+	CacheCluster = ""
+	// Runtime sets the WASM runtime to use for extensions
+	Runtime = defaultRuntime
+
+	defaultConfigSource = &envoy_config_core_v3.ConfigSource{
+		ConfigSourceSpecifier: &envoy_config_core_v3.ConfigSource_Ads{
+			Ads: &envoy_config_core_v3.AggregatedConfigSource{},
+		},
+		ResourceApiVersion: envoy_config_core_v3.ApiVersion_V3,
+	}
+)
+
+// AddWasmPluginsToListener adds WasmPlugins to listener filterChains
+func AddWasmPluginsToListeners(
+	listeners []*xdslistener.Listener,
+	pluginsMap map[networking.EnvoyFilter_Patch_FilterClass][]*model.WasmPluginWrapper,
+	proxy *model.Proxy,
+	push *model.PushContext,
+) []*xdslistener.Listener {
+	ret := []*xdslistener.Listener{}
+	for _, listener := range listeners {
+		ret = append(ret, AddWasmPluginsToListener(listener, pluginsMap, proxy, push))
+	}
+	return ret
+}
+
+// AddWasmPluginsToListener adds WasmPlugins to listener filterChains
+func AddWasmPluginsToListener(
+	listener *xdslistener.Listener,
+	pluginsMap map[networking.EnvoyFilter_Patch_FilterClass][]*model.WasmPluginWrapper,
+	proxy *model.Proxy,
+	push *model.PushContext,
+) *xdslistener.Listener {
+	if listener == nil {
+		return nil
+	}
+
+	for fcIndex, fc := range listener.FilterChains {
+		// copy extensions map
+		extensions := make(map[networking.EnvoyFilter_Patch_FilterClass][]*model.WasmPluginWrapper)
+		for phase, list := range pluginsMap {
+			extensions[phase] = []*model.WasmPluginWrapper{}
+			extensions[phase] = append(extensions[phase], list...)
+		}
+		hcm := &hcm_filter.HttpConnectionManager{}
+		hcmIndex := -1
+		for i, f := range fc.Filters {
+			if f.Name == wellknown.HTTPConnectionManager && f.GetTypedConfig() != nil {
```

Comment:
This is fine for a POC but for the real implementation we need to hook deeper into the LDS generation to avoid double marshalling

Reply:
I moved it to right after plugin gen now